File: src/test/java/ru/vito/desktop/app/service/impl/DefinerTest.java
package ru.vito.desktop.app.service.impl;

import org.junit.jupiter.api.Test;
import ru.vito.desktop.app.models.DefinedUsers;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class DefinerTest {

    private Definer definer = new Definer();

    @Test
    void givenNullRequest_Ok() {
        // ACT
        final DefinedUsers response = definer.define(null);

        // ASSERT
        assertNull(response);
    }

    @Test
    void givenEmptyRequest_Ok() {
        // ACT
        final DefinedUsers response = definer.define(new HashMap<>());

        // ASSERT
        assertNotNull(response);
        assertNotNull(response.getEmailToLoginMap());
        assertEquals(0, response.getEmailToLoginMap().size());
    }

    @Test
    void givenSingleElementRequest_Ok() {
        // ARRANGE
        final Map<String, HashSet<String>> request = new HashMap<>();
        final List<String> requestEmailsList = Arrays.asList("xxx@ya.ru", "foo@gmail.com", "lol@mail.ru");
        final HashSet<String> requestEmailsSet = new HashSet<>(requestEmailsList);
        request.put("user1 ", requestEmailsSet);

        // ACT
        final DefinedUsers response = definer.define(request);

        // ASSERT
        assertNotNull(response);
        assertEquals(3, response.getEmailToLoginMap().size());
        assertEquals(1, new HashSet<>(response.getEmailToLoginMap().values()).size());
        final Set<String> emailsSet = response.getEmailToLoginMap().keySet();
        assertEquals(3, emailsSet.size());
        assertTrue(emailsSet.containsAll(requestEmailsList));
    }

    @Test
    void givenSingleMergedElementRequest_Ok() {
        // ARRANGE
        final Map<String, HashSet<String>> request = new HashMap<>();
        final List<String> requestEmailsList = Arrays.asList("xxx@ya.ru", "xxx@ya.ru", "lol@mail.ru");
        final HashSet<String> requestEmailsSet = new HashSet<>(requestEmailsList);
        request.put("user1 ", requestEmailsSet);

        // ACT
        final DefinedUsers response = definer.define(request);

        // ASSERT
        assertNotNull(response);
        assertEquals(2, response.getEmailToLoginMap().size());
        assertEquals(1, new HashSet<>(response.getEmailToLoginMap().values()).size());
        final Set<String> emailsSet = response.getEmailToLoginMap().keySet();
        assertEquals(2, emailsSet.size());
        assertTrue(emailsSet.containsAll(requestEmailsList));
    }

    @Test
    void givenFewElementsList_Ok() {
        // ARRANGE
        final Map<String, HashSet<String>> request = new HashMap<>();
        request.put("user1 ", new HashSet<>(Arrays.asList("xxx@ya.ru", "foo@gmail.com", "lol@mail.ru")));
        request.put("user2 ", new HashSet<>(Collections.singletonList("ups@pisem.net")));

        // ACT
        final DefinedUsers response = definer.define(request);

        // ASSERT
        assertNotNull(response);
        assertEquals(4, response.getEmailToLoginMap().size());
        assertEquals(2, new HashSet<>(response.getEmailToLoginMap().values()).size());
        final Set<String> emailsSet = response.getEmailToLoginMap().keySet();
        assertEquals(4, emailsSet.size());
    }

    @Test
    void givenFewElementsList_mergeKit_Ok() {
        // ARRANGE
        final Map<String, HashSet<String>> request = new HashMap<>();
        request.put("user1 ", new HashSet<>(Arrays.asList("xxx@ya.ru", "foo@gmail.com", "lol@mail.ru")));
        request.put("user2 ", new HashSet<>(Arrays.asList("foo@gmail.com", "ups@pisem.net")));
        request.put("user3 ", new HashSet<>(Arrays.asList("xyz@pisem.net", "vasya@pupkin.com")));
        request.put("user4 ", new HashSet<>(Arrays.asList("ups@pisem.net", "aaa@bbb.ru")));
        request.put("user5 ", new HashSet<>(Collections.singletonList("xyz@pisem.net")));

        // ACT
        final DefinedUsers response = definer.define(request);

        // ASSERT
        assertNotNull(response);
        assertEquals(7, response.getEmailToLoginMap().size());
        assertEquals(2, new HashSet<>(response.getEmailToLoginMap().values()).size());
        final Set<String> emailsSet = response.getEmailToLoginMap().keySet();
        assertEquals(7, emailsSet.size());
    }

    @Test
    void givenFewElementsList_hardKit_Ok() {
        // ARRANGE
        final Map<String, HashSet<String>> request = new HashMap<>();
        request.put("user1 ", new HashSet<>(Arrays.asList("e1@ya.ru", "e2@ya.ru")));
        request.put("user2 ", new HashSet<>(Arrays.asList("e3@ya.ru", "e4@ya.ru")));
        request.put("user3 ", new HashSet<>(Arrays.asList("e1@ya.ru", "e3@ya.ru")));

        // ACT
        final DefinedUsers response = definer.define(request);

        // ASSERT
        assertNotNull(response);
        assertEquals(4, response.getEmailToLoginMap().size());
        assertEquals(1, new HashSet<>(response.getEmailToLoginMap().values()).size());
        final Set<String> emailsSet = response.getEmailToLoginMap().keySet();
        assertEquals(4, emailsSet.size());
    }
}

File: src/test/kotlin/ru/vito/desktop/app/service/impl/DefinerTest.kt
package ru.vito.desktop.app.service.impl

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import ru.vito.desktop.app.models.DefinedUsers
import java.util.*

internal class DefinerTest {

    private val definer = Definer()

    @Test
    fun givenNullRequest_Ok() {
        // ACT
        val response = definer.define(Collections.emptyMap())

        // ASSERT
        assertNotNull(response) // TODO to valid test
        assertEquals(0, response.emailToLoginMap.size)
    }

    @Test
    fun givenEmptyRequest_Ok() {
        // ACT
        val response : DefinedUsers = definer.define(HashMap())

        // ASSERT
        assertNotNull(response)
        assertNotNull(response.emailToLoginMap)
        assertEquals(0, response.emailToLoginMap.size)
    }

    @Test
    fun givenSingleElementRequest_Ok() {
        // ARRANGE
        val request = HashMap<String, HashSet<String>>()
        val requestEmailsList = Arrays.asList("xxx@ya.ru", "foo@gmail.com", "lol@mail.ru")
        val requestEmailsSet = HashSet(requestEmailsList)
        request["user1 "] = requestEmailsSet

        // ACT
        val response = definer.define(request)

        // ASSERT
        assertNotNull(response)
        assertEquals(3, response.emailToLoginMap.size)
        assertEquals(1, HashSet(response.emailToLoginMap.values).size)
        val emailsSet = response.emailToLoginMap.keys
        assertEquals(3, emailsSet.size)
        assertTrue(emailsSet.containsAll(requestEmailsList))
    }

    @Test
    fun givenSingleMergedElementRequest_Ok() {
        // ARRANGE
        val request = HashMap<String, HashSet<String>>()
        val requestEmailsList = Arrays.asList("xxx@ya.ru", "xxx@ya.ru", "lol@mail.ru")
        val requestEmailsSet = HashSet(requestEmailsList)
        request["user1 "] = requestEmailsSet

        // ACT
        val response = definer.define(request)

        // ASSERT
        assertNotNull(response)
        assertEquals(2, response.emailToLoginMap.size)
        assertEquals(1, HashSet(response.emailToLoginMap.values).size)
        val emailsSet = response.emailToLoginMap.keys
        assertEquals(2, emailsSet.size)
        assertTrue(emailsSet.containsAll(requestEmailsList))
    }

    @Test
    fun givenFewElementsList_Ok() {
        // ARRANGE
        val request = HashMap<String, HashSet<String>>()
        request["user1 "] = HashSet(Arrays.asList("xxx@ya.ru", "foo@gmail.com", "lol@mail.ru"))
        request["user2 "] = HashSet(listOf("ups@pisem.net"))

        // ACT
        val response = definer.define(request)

        // ASSERT
        assertNotNull(response)
        assertEquals(4, response.emailToLoginMap.size)
        assertEquals(2, HashSet(response.emailToLoginMap.values).size)
        val emailsSet = response.emailToLoginMap.keys
        assertEquals(4, emailsSet.size)
    }

    @Test
    fun givenFewElementsList_mergeKit_Ok() {
        // ARRANGE
        val request = HashMap<String, HashSet<String>>()
        request["user1 "] = HashSet(Arrays.asList("xxx@ya.ru", "foo@gmail.com", "lol@mail.ru"))
        request["user2 "] = HashSet(Arrays.asList("foo@gmail.com", "ups@pisem.net"))
        request["user3 "] = HashSet(Arrays.asList("xyz@pisem.net", "vasya@pupkin.com"))
        request["user4 "] = HashSet(Arrays.asList("ups@pisem.net", "aaa@bbb.ru"))
        request["user5 "] = HashSet<String>(listOf<String>("xyz@pisem.net"))

        // ACT
        val response = definer.define(request)

        // ASSERT
        assertNotNull(response)
        assertEquals(7, response.emailToLoginMap.size)
        assertEquals(2, HashSet(response.emailToLoginMap.values).size)
        val emailsSet = response.emailToLoginMap.keys
        assertEquals(7, emailsSet.size)
    }

    @Test
    fun givenFewElementsList_hardKit_Ok() {
        // ARRANGE
        val request = HashMap<String, HashSet<String>>()
        request["user1 "] = HashSet(Arrays.asList("e1@ya.ru", "e2@ya.ru"))
        request["user2 "] = HashSet(Arrays.asList("e3@ya.ru", "e4@ya.ru"))
        request["user3 "] = HashSet(Arrays.asList("e1@ya.ru", "e3@ya.ru"))

        // ACT
        val response = definer.define(request)

        // ASSERT
        assertNotNull(response)
        assertEquals(4, response.emailToLoginMap.size)
        assertEquals(1, HashSet(response.emailToLoginMap.values).size)
        val emailsSet = response.emailToLoginMap.keys
        assertEquals(4, emailsSet.size)
    }
}